### PR TITLE
USWDS - POAM: August '24

### PR DESCRIPTION
**Pull request by mahoneycm**
# Summary

Resolved dependency vulnerabilities via `npm audit fix` and installed the USWDS Compile's new version.

## Breaking change

This is not a breaking change.

## Related issue

[USWDS-Team - POAM: August 2024](https://github.com/uswds/uswds-team/issues/370)
Closes https://github.com/uswds/uswds/security/dependabot/81

## Preview link

[Preview link →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-POAM-july-2024/)

## Major changes

**Vulnerabilities before update**

```node
61 vulnerabilities (5 low, 11 moderate, 45 high)
```

**After update**
```node
60 vulnerabilities (5 low, 11 moderate, 44 high)
```

> [!NOTE]
> This PR also restores Prettier HTML behaviors that format our HTML-Templates for Site

## Testing and review

1. Pull down and run `npm install`
2. Confirm no installation errors
3. Run Compile scripts
4. Confirm no build errors
5. Run `npm run test`
6. Confirm no testing errors
7. Run `npm run lint`
8. Confirm linter runs without error
9. Start local or visit preview link
10. Confirm there are no visual regressions

## Dependency updates

| Dependency name | Previous version | Updated version |
| --- | --- | --- |
| @18f/identity-stylelint-config | 4.0.0 | 4.1.0 |
| @babel/core | 7.24.5 | 7.25.2 |
| @babel/preset-env | 7.24.5 |  7.25.3 |
| @types/node | 20.12.11 | 20.14.10 |
| autoprefixer | 10.4.19 | 10.4.20 |
| axe-core | 4.9.1 | 4.10.0 |
| mocha | 10.4.0 | 10.7.0 |
| postcss | 8.4.38 | 8.4.40 |
| postcss-preset-env | 9.5.11 | 9.6.0 |
| prettier | 3.2.5 | 3.3.3 |
| sass | 1.77.0 | 1.77.6 |
| sass-embedded | 1.77.0 | 1.77.5 |
| snyk | 1.1291.0 | 1.1292.2 |
| stylelint | 16.5.0 | 16.8.1 |
| typescript | 5.4.5 | 5.5.4 |
| webpack | 5.91.0 | 5.93.0 |

### CODE REVIEW

File: package.json
Comment (mejiaj) [inline package.json:1]: **polish**: There are additional patch and minor updates we can make.

**Patch**
- `@babel/core` 7.24.7  →  7.24.9
- `@babel/preset-env` 7.24.7  →  7.24.8
- `prettier` 3.3.2  →   3.3.3
- `sass` 1.77.6  →  1.77.8
- `sass-embedded` 1.77.5  →  1.77.8

**Minor**
- `stylelint` 16.6.1  →  16.7.0
- `webpack` 5.92.1  →  5.93.0

**Review by mejiaj — COMMENTED** [review]
Thanks @mahoneycm, found additional non-breaking updates to dependencies we can make. Otherwise LGTM.

**Tested the following**
- [x] Confirming `56` vulnerabilities before, and `55` after.
- [x] Dependency table in PR description is accurate.
- [ ] There aren't further dependency updates to be made.

**Comment by mahoneycm** [comment]
@mejiaj Updating Sass brings on a new Deprecation warning on build:

> Sass's behavior for declarations that appear after nested
rules will be changing to match the behavior specified by CSS in an upcoming
version. To keep the existing behavior, move the declaration above the nested
rule. To opt into the new behavior, wrap the declaration in `& {}`

It sites this in almost 600 locations. The notification also displays when installing on Site and running the build command. Would we want to move forward with this change and create a new issue to address? Or should we leave the current version and address it without the build warnings? Since it is not a vulnerability fix, I think we should be fine holding off while we address the future deprecation

**Comment by mahoneycm** [comment]
@mejiaj Sounds good! The other dependency updates were updated without a problem 👍 

Created #5980 to track this issue and discussion.

**Comment by amyleadem** [comment]
> [!Note]
> Added this to the 3.9.0 milestone to make sure it gets included.

File: package.json
Comment (mejiaj) [reply package.json:1]: Updated in ef07fbd, except SASS due to deprecation warnings.

**Review by mejiaj — APPROVED** [review]
Looks good. Confirming `npm audit` report and dependency table is accurate.

**Review by amyleadem — COMMENTED** [review]
Looking good. Thanks for this, @mahoneycm.

Everything is working well, but I noticed an error when running `npm run lint`:
```
[06:58:14] 'lintSass' errored after 9.9 s
[06:58:14] Error: 
packages/uswds-core/src/styles/functions/units/units.scss
  13:46  ✖  Unexpected extension ".scss" in @use  scss/load-partial-extension

✖ 1 problem (1 error, 0 warnings)

    at lintSass **(/Users/amy.leadem/Dev/github/workspace/uswds/tasks/lint.js:52:22)**
```

However, it looks like it  is pre-existing in `develop`. Wondering if everyone would be ok if we snuck the fix for that in here or if we should break out a new issue and PR.

**Comment by mejiaj** [comment]
@mahoneycm moving to `in progress` to run requested changes from Amy.

**Comment by mahoneycm** [comment]
@amyleadem  I was unable to replicate the error. For some reason, the linter did not return the error for me on either this branch or `develop`. Would that be any cause for concern worth investigating?

image

In the meantime, I went ahead and removed the `.scss`extention in the mentioned file. I tested by updating units and setting it to a non-unit value. The `error-not-token` error ran as expected 👍

**Review by amyleadem — APPROVED** [review]
Looks good! Thanks for that fix. Let's keep checking in to see if we continue to get different results when linting. I don't think any action is needed at this time.

**Comment by mahoneycm** [comment]
## August updates:

```
@babel/core          7.24.9  →    7.25.2
@babel/preset-env    7.24.8  →    7.25.3
autoprefixer        10.4.19  →   10.4.20
axe-core              4.9.1  →    4.10.0
mocha                10.6.0  →    10.7.0
postcss              8.4.39  →    8.4.40
snyk               1.1292.1  →  1.1292.2
stylelint            16.7.0  →    16.8.1
typescript            5.5.3  →     5.5.4
```

**Review by amyleadem — APPROVED** [review]
Looks good to me! 

> [!Note]
> I made a few small updates to the dependency table in the PR description to match the changes I see in package.json

File: package.json [hunk 1]
```diff
@@ -155,34 +155,34 @@
...
+    "sass": "1.77.6",
+    "sass-embedded": "1.77.5",
     "sass-loader": "13.3.2",
     "sass-true": "6.0.1",
     "sinon": "12.0.1",
-    "snyk": "1.1291.0",
+    "snyk": "1.1292.2",
     "style-loader": "3.3.3",
-    "stylelint": "16.5.0",
+    "stylelint": "16.8.1",
```
Comment (mejiaj) [inline package.json:178]: **issue**: This task is failing locally.

```sh
npm run lint:sass
```

Results in an error:
```sh

> @uswds/uswds@3.8.1 lint:sass
> gulp lintSass

[15:16:42] Using gulpfile ~/web/uswds/gulpfile.js
[15:16:42] Starting 'lintSass'...
(node:13738) [stylelint:002] DeprecationWarning: The CommonJS Node.js API is deprecated.
See https://stylelint.io/migration-guide/to-16
(Use `node --trace-deprecation ...` to show where the warning was created)
[15:16:53] Finished 'lintSass' after 11 s
```

---

This also appears, but doesn't block `npx gulp buildSass`.

File: package.json [hunk 2]
```diff
@@ -155,34 +155,34 @@
...
-    "postcss": "8.4.38",
+    "postcss": "8.4.40",
     "postcss-csso": "6.0.1",
     "postcss-discard-comments": "6.0.2",
     "postcss-import": "15.1.0",
     "postcss-loader": "7.3.3",
-    "postcss-preset-env": "9.5.11",
-    "prettier": "3.2.5",
+    "postcss-preset-env": "9.6.0",
+    "prettier": "3.3.3",
```
Comment (mejiaj) [inline package.json:168]: **issue**: This is failing silently for `npm run prettier:html`.

```sh

> @uswds/uswds@3.8.1 prettier:html
> npx prettier --write './html-templates/**/**/*.html'

```

Seeing the same in `develop` branch too.

image

**Review by mejiaj — COMMENTED** [review]
@mahoneycm thanks for updating this.

I've run `npm audit` on clean installs for `develop` and this branch. Shows less than the PR description & no difference from either branch.

| Before | After |
| :----- | :---- |
| `56 vulnerabilities (1 low, 11 moderate, 44 high)` |`56 vulnerabilities (1 low, 11 moderate, 44 high)`|

### Things I tested

- `npm run lint:sass` [^1]
- `npm run lint:js`
- `npm run prettier`
- `npm run test:unit`
- `npm run build:html` [^1]

[^1]: Error reported - additional info in PR comment.

File: package.json (hunk 1 above)
Comment (mahoneycm) [reply package.json:178]: **note:** This is just a deprecation _warning_. It was introduced in #5782 and flagged in [this comment](https://github.com/uswds/uswds/pull/5782#pullrequestreview-2043394195)

> **Stylelint CommonJS Node.js API is deprecated**
> This isn't an issue for this version of Stylelint (`16.5.0`) but will prevent us from updating to the next major version unless we migrate to ESM. [^1]

[^1]: [Stylint: Deprecated CommonJS API](https://stylelint.io/migration-guide/to-16/#deprecated-commonjs-api)

If you introduce a sass lint error it'll appropriately flag it.

image

Comment (mejiaj) [reply package.json:178]: Thanks @mahoneycm. Is there an issue for it so we can triage? Seems urgent.

File: package.json (hunk 2 above)
Comment (mahoneycm) [reply package.json:167]: GREAT catch. Looks like starting with Prettier `3.x.x`, prettier uses our `.gitignore` to decide what files to ignore as well.

[Prettier changelog note about this and how to replicate previous behavior →](https://prettier.io/blog/2023/07/05/3.0.0.html#ignore-gitignored-files-by-default-14731httpsgithubcomprettierprettierpull14731-by-fiskerhttpsgithubcomfisker)

I added the `--ignore-path=.prettierignore` flag to our prettier scripts to maintain our previous ignores. Luckily, we don't have any inadvertent changes. The script now appropriately formats the `html-templates`

```diff
- "prettier:html": "npx prettier --write './html-templates/**/**/*.html'",
- "prettier:js": "npx prettier --write './packages/**/**/*.js'",
- "prettier:sass": "npx prettier --write './packages/**/**/*.scss'",
+ "prettier:html": "npx prettier --write './html-templates/**/**/*.html' --ignore-path=.prettierignore",
+ "prettier:js": "npx prettier --write './packages/**/**/*.js' --ignore-path=.prettierignore",
+ "prettier:sass": "npx prettier --write './packages/**/**/*.scss' --ignore-path=.prettierignore",
```

Resolved in [74e8c59](https://github.com/uswds/uswds/pull/5972/commits/74e8c597d2fd8705ac7e945699b5bf1fb6ac022b)

File: package.json (hunk 1 above)
Comment (mahoneycm) [reply package.json:177]: We have #4986 which is broken into sub-tasks. I've added [a comment](https://github.com/uswds/uswds/issues/4986#issuecomment-2299030609) outlining the need to migrate to ESM before the next major version of Stylelint is released.

I'll add a discussion item to a team sync and we can discuss the ESM migration and where it fits into our roadmap.

**Review by mejiaj — APPROVED** [review]
Thanks for the updates. I was able to successfully run prettier on HTML templates.

image

**Review by thisisdano — APPROVED** [review]

**Pull request merged**